File: internal/workflow/workflow.go
package workflow

import (
	"fmt"

	"imagecompression/internal/compress"
	"imagecompression/internal/prepare"
	"imagecompression/internal/upload"
)

type Options struct {
	PrepareOperations []prepare.Operation
	PrepareOverwrite  bool
	CompressOptions   compress.BatchOptions
}

type Result struct {
	Prepare  prepare.Result       `json:"prepare"`
	Compress compress.BatchResult `json:"compress"`
	Upload   upload.Result        `json:"upload"`
}

type Hooks struct {
	Prepare  func([]prepare.Operation, bool) (prepare.Result, error)
	Compress func(compress.BatchOptions) (compress.BatchResult, error)
	Upload   func(string) (upload.Result, error)
}

func RunPrepareCompressUpload(options Options, hooks Hooks) (Result, error) {
	if hooks.Prepare == nil || hooks.Compress == nil || hooks.Upload == nil {
		return Result{}, fmt.Errorf("workflow hooks 未配置完整")
	}

	prepareResult, err := hooks.Prepare(options.PrepareOperations, options.PrepareOverwrite)
	if err != nil {
		return Result{}, err
	}

	compressOptions := options.CompressOptions
	compressOptions.InputDir = prepareResult.OutputDir
	compressResult, err := hooks.Compress(compressOptions)
	if err != nil {
		return Result{Prepare: prepareResult}, err
	}

	uploadResult, err := hooks.Upload(compressResult.OutputDir)
	if err != nil {
		return Result{Prepare: prepareResult, Compress: compressResult}, err
	}

	return Result{
		Prepare:  prepareResult,
		Compress: compressResult,
		Upload:   uploadResult,
	}, nil
}

File: src/core/workflow.rs
use crate::config;
use crate::core::{compress, prepare, upload as up};

pub struct Options {
    pub prepare_ops: Vec<prepare::Operation>,
    pub prepare_overwrite: bool,
    pub compress_opts: compress::BatchOptions,
    pub upload_cfg: Option<config::UploadConfig>,
    pub do_upload: bool,
}

pub struct WorkflowResult {
    pub prepare: prepare::PrepareResult,
    pub compress: compress::BatchResult,
    pub upload: Option<up::UploadResult>,
}

pub fn run_prepare_compress_upload(opts: Options) -> anyhow::Result<WorkflowResult> {
    let prep = prepare::execute_operations(opts.prepare_ops, opts.prepare_overwrite, None)?;

    let mut comp_opts = opts.compress_opts;
    comp_opts.input_dir = prep.output_dir.clone();

    let comp = compress::compress_directory(std::sync::Arc::new(()), comp_opts, None, None)?;

    let up_res = if opts.do_upload {
        if let Some(uc) = opts.upload_cfg {
            let eff = up::effective_config(uc, &comp.output_dir);
            let mut u = crate::core::upload::build_uploader(eff);
            Some(up::upload_directory(
                &mut *u,
                &comp.output_dir,
                up::Options { recursive: true },
                None,
            )?)
        } else {
            None
        }
    } else {
        None
    };

    Ok(WorkflowResult {
        prepare: prep,
        compress: comp,
        upload: up_res,
    })
}
